pkg/util/containers/v2/metrics/collector_containerd.go

```diff
@@ -0,0 +1,416 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// +build containerd
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	wstats "github.com/Microsoft/hcsshim/cmd/containerd-shim-runhcs-v1/stats"
+	v1 "github.com/containerd/cgroups/stats/v1"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/typeurl"
+	"github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/providers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/system"
+)
+
+const (
+	containerdCollectorID    = "containerd"
+	containerdCacheKeyPrefix = "containerd-stats-"
+	containerdCacheTTL       = 10 * time.Second
+)
+
+func init() {
+	metricsProvider.registerCollector(collectorMetadata{
+		id:       containerdCollectorID,
+		priority: 1, // Less than the "system" collector, so we can rely on cgroups directly if possible
+		runtimes: []string{RuntimeNameContainerd},
+		factory: func() (Collector, error) {
+			return newContainerdCollector()
+		},
+	})
+}
+
+type containerdCollector struct {
+	client         cutil.ContainerdItf
+	lastScrapeTime time.Time
+}
+
+func newContainerdCollector() (*containerdCollector, error) {
+	client, err := cutil.GetContainerdUtil()
+	if err != nil {
+		return nil, ErrNothingYet
```

Comment:
shouldn't this be based on the error type returned by `GetContainerdUtil` ? Thinking mainly about permanent failure

Reply:
Changed to `return nil, ErrPermaFail` as discussed offline.

Reply:
It should check if `containerd` feature is present as well or not

Reply:
I've added the check @vboulineau